### PR TITLE
huge overhaul/update of cascade classifier training

### LINKED ISSUE #6400: Location of expanded documentation for opencv_annotation, opencv_visualisation and opencv_traincascade
I would like to request where the extra information should go. I would be eager to open up some wiki pages, due to the very ease of placing and editing stuff there. On the other hand, you guys might prefer that documentation goes inside the actual docs, and thus I am wondering where I would be adding those guides then?

@mshabunin or @alalek any guidelines?

### PR DESCRIPTION
**Pull request by StevenPuttemans**
### This pullrequest changes

Complete overhaul of the cascade classifier training tutorial in documentation. Also added documentation for the OpenCV annotation and visualisation tool.

resolves #6400

### CODE REVIEW

**Comment by StevenPuttemans** [comment]
Build succeeded, but going to apply some fixes of visuals first! Will report back when finished!

Current tutorial: http://pullrequest.opencv.org/buildbot/export/pr/7112/docs/dc/d88/tutorial_traincascade.html

File: doc/tutorials/objdetect/traincascade.markdown [hunk 1]
```diff
@@ -4,66 +4,28 @@ Cascade Classifier Training {#tutorial_traincascade}
...
-training size.
+Negative samples are taken from arbitrary images, not containing objects you want to detect. These negative images, from which the samples are generated, should be listed in a special negative image file, which should be created manually. It is a text file in which each line contains an the location of such a negative image *(absolute paths ensure that everything works as expected)*. Note that negative samples and sample images are also called background samples or background images, and are used interchangeably in this document.
 
-An example of description file:
+Described images may be of different sizes. However, each image should be equal or larger than the desired training window size *(which corresponds to the model dimensions, most of the times being the average size of your object)*, because these images are used to subsample a given negative image into several image samples having this training window size.
+
+An example of such a negative description file:
 
 Directory structure:
 @code{.text}
```
Comment (savuor) [inline doc/tutorials/objdetect/traincascade.markdown]: This text is displayed incorrectly

**Comment by StevenPuttemans** [comment]
The now submitted and being build form should be final. It seems that markdown and doxygen only work decently together in very strict circumstances. Finally I just used html tags to apply some local italic and bold iterations that did not combine well inside bulletpoints (which seem to be a known limitation of markup inside doxygen)

**Comment by mshabunin** [comment]
@StevenPuttemans , you can use `-` to mark list items and `_word_` or `__word__` to format words italic or bold. This variant should work in most cases.

I suggest replacing multiple `[opencv\_traincascade](...)` links with plain text and adding single source link in the beginning.

File: doc/tutorials/objdetect/traincascade.markdown [hunk 2]
```diff
@@ -4,66 +4,28 @@ Cascade Classifier Training {#tutorial_traincascade}
...
-is created using opencv_createsamples utility.
+ * If you come accross any tutorial mentioning the old opencv\_haartraining tool <i>(which is depricated and still using the OpenCV1.x interface)</i>, then please ignore that tutorial and stick to the [opencv\_traincascade](https://github.com/opencv/opencv/tree/master/apps/traincascade) tool. This tool is a is a newer version, written in C++ in accordance to the OpenCV 2.x and OpenCV 3.x API. The [opencv\_traincascade](https://github.com/opencv/opencv/tree/master/apps/traincascade) supports both HAAR like wavelet features @cite Viola01 and LBP (Local Binary Patterns) @cite Liao2007 features. LBP features yield integer precision in contrast to HAAR features, yielding floating point precision, so both training and detection with LBP are several times faster then with HAAR features. Regarding the LBP and HAAR detection quality, it mainly depends on the training data used and the training parameters selected.It's possible to train a LBP-based classifier that will provide almost the same quality as HAAR-based one, within a percentage of the training time.
```
Comment (mshabunin) [inline doc/tutorials/objdetect/traincascade.markdown]: No space after punctuation: `selected.It's`

Comment (mshabunin) [inline doc/tutorials/objdetect/traincascade.markdown]: Should be `deprecated`: `depricated `

Comment (mshabunin) [inline doc/tutorials/objdetect/traincascade.markdown]: Repeated: `is a is a`

File: doc/tutorials/objdetect/traincascade.markdown [hunk 3]
```diff
@@ -4,66 +4,28 @@ Cascade Classifier Training {#tutorial_traincascade}
...
+
+ * The newer cascade classifier detection interface from OpenCV 2.x and OpenCV 3.x (see CascadeClassifier class in objdetect module) supports working with both old and new model formats. [opencv\_traincascade](https://github.com/opencv/opencv/tree/master/apps/traincascade) can even save (export) a trained cascade in the older format if for some reason you are stuck using the old interface. At least training the model could then be done in the most stable interface.
```
Comment (mshabunin) [inline doc/tutorials/objdetect/traincascade.markdown]: Will be better to add reference to class (`@ref CascadeClassifier`): `see CascadeClassifier class`

File: doc/tutorials/objdetect/traincascade.markdown [hunk 4]
```diff
@@ -4,66 +4,28 @@ Cascade Classifier Training {#tutorial_traincascade}
...
+
+ * The [opencv\_traincascade](https://github.com/opencv/opencv/tree/master/apps/traincascade) application can use TBB for multi-threading. To use it in multicore mode OpenCV must be built with TBB support enabled.
+
+Preparation of the training data
+--------------------------------
+
+For training a boosted cascade of weak classifiers we need a set of positive samples (containing actual objects you want to detect) and a set of negative images (containing everything you do not want to detect). The set of negative samples must be prepared manually, whereas set of positive samples is created using the [opencv\_traincascade](https://github.com/opencv/opencv/tree/master/apps/createsamples) application.
```
Comment (mshabunin) [inline doc/tutorials/objdetect/traincascade.markdown]: Should be `create_samples`: `[opencv\_traincascade](https://github.com/opencv/opencv/tree/master/apps/createsamples)`

File: doc/tutorials/objdetect/traincascade.markdown [hunk 5]
```diff
@@ -4,66 +4,28 @@ Cascade Classifier Training {#tutorial_traincascade}
...
-Negative samples are enumerated in a special file. It is a text file in which each line contains an
-image filename (relative to the directory of the description file) of negative sample image. This
-file must be created manually. Note that negative samples and sample images are also called
-background samples or background images, and are used interchangeably in this document.
-Described images may be of different sizes. But each image should be (but not nessesarily) larger
-than a training window size, because these images are used to subsample negative image to the
-training size.
+Negative samples are taken from arbitrary images, not containing objects you want to detect. These negative images, from which the samples are generated, should be listed in a special negative image file, which should be created manually. It is a text file in which each line contains an the location of such a negative image <i>(absolute paths ensure that everything works as expected)</i>. Note that negative samples and sample images are also called background samples or background images, and are used interchangeably in this document.
```
Comment (mshabunin) [inline doc/tutorials/objdetect/traincascade.markdown]: I'd rephrase it: 
```
...should be listed in a special negative image file. It is a text file in which each line contains an the location of such a negative image <i>(absolute paths ensure that everything works as expected)</i>.
```
 -> 
```
...should be listed in a special negative image file one path per line (can be absolute or relative).
```

**Comment by StevenPuttemans** [comment]
@mshabunin thanks for all the pointers, will fix everything as to your suggestions!

File: doc/tutorials/objdetect/traincascade.markdown [hunk 6]
```diff
@@ -184,146 +103,115 @@ img/img2.jpg  2  100 200 50 50   50 30 25 25
...
-
-        This argument is used to determine how precise your model should keep learning and when to stop.
-        A good guideline is to train not further than 10e-5, to ensure the model does not overtrain on your training data.
-        By default this value is set to -1 to disable this feature.
-
--#  Cascade parameters:
-
-    -   -stageType \<BOOST(default)\>
+@code{.text}
+opencv\_annotation --annotations=/path/to/annotations/file.txt --images=/path/to/image/folder/`
```
Comment (mshabunin) [inline doc/tutorials/objdetect/traincascade.markdown]: Let's replace all `opencv\_` with `opencv_`. I think it will work fine.
On this line there is backtick at the end.

File: doc/tutorials/objdetect/traincascade.markdown [hunk 7]
```diff
@@ -184,146 +103,115 @@ img/img2.jpg  2  100 200 50 50   50 30 25 25
...
+  - `-numThreads <max_number_of_threads>` : Maximum number of threads to use during training. Notice that the actual number of used threads may be lower, depending on your machine and compilation options. By default, the maximum available threads are selected if you built OpenCV with TBB support, which is needed for this optimization.
+  - `-acceptanceRatioBreakValue <break_value>` : This argument is used to determine how precise your model should keep learning and when to stop. A good guideline is to train not further than 10e-5, to ensure the model does not overtrain on your training data. By default this value is set to -1 to disable this feature.
 
-        Type of boosted classifiers: DAB - Discrete AdaBoost, RAB - Real AdaBoost, LB - LogitBoost,
-        GAB - Gentle AdaBoost.
+- Cascade parameters:
+  - `-stageType <BOOST(default)>` : Type of stages. Only boosted classifier are supported as a stage type at the moment.
```
Comment (mshabunin) [inline doc/tutorials/objdetect/traincascade.markdown]: Should be `... classifiers are supported ...`

File: doc/tutorials/objdetect/traincascade.markdown [hunk 8]
```diff
@@ -184,146 +103,115 @@ img/img2.jpg  2  100 200 50 50   50 30 25 25
...
-    -   -mode \<BASIC (default) | CORE | ALL\>
+Example of the HAAR/LBP face model ran on a given window of Angelina Jolie, which had the same preprocessing as cascade classifier files -->24x24 pixel image, grayscale conversion and histogram equalisation:
 
-        Selects the type of Haar features set used in training. BASIC use only upright features,
-        while ALL uses the full set of upright and 45 degree rotated feature set. See @cite Lienhart02
-        for more details.
+_A video is made with for each stage each feature visualised:_
 
--#  Local Binary Patterns parameters:
+![9757_05_14](https://cloud.githubusercontent.com/assets/4621239/14203748/1bef5efc-f800-11e5-84b2-939d47e46161.png)
```
Comment (mshabunin) [inline doc/tutorials/objdetect/traincascade.markdown]: Images should be added to the repository, for example to the `opencv/doc/tutorials/objdetect/images/` folder.

File: doc/tutorials/objdetect/traincascade.markdown [hunk 9]
```diff
@@ -4,66 +4,30 @@ Cascade Classifier Training {#tutorial_traincascade}
...
+
+ - The newer cascade classifier detection interface from OpenCV 2.x and OpenCV 3.x (@ref CascadeClassifier) supports working with both old and new model formats. opencv\_traincascade can even save (export) a trained cascade in the older format if for some reason you are stuck using the old interface. At least training the model could then be done in the most stable interface.
```
Comment (mshabunin) [inline doc/tutorials/objdetect/traincascade.markdown]: Doxygen produces a warning here, let's try to use `@ref cv::CascadeClassifier`.

**Comment by mshabunin** [comment]
@StevenPuttemans , I have no other comments.

**Comment by StevenPuttemans** [comment]
@mshabunin I applied all suggested fixes. Also noticed that 

`doc/tutorials/objdetect/images/Cascade_Classifier_Tutorial_Cover.jpg`

is actually redundant for 

`doc/tutorials/objdetect/cascade_classifier/images/Cascade_Classifier_Tutorial_Cover.jpg`

and can thus be removed without issues.

**Comment by mshabunin** [comment]
@StevenPuttemans , looks like cover image is broken. Please revert it:
```
git checkout improve_cascadeclassifier_training_tutorial
git checkout master -- doc/tutorials/objdetect/images/Cascade_Classifier_Tutorial_Cover.jpg
git commit --amend -- doc/tutorials/objdetect/images/Cascade_Classifier_Tutorial_Cover.jpg
```

**Comment by StevenPuttemans** [comment]
Fixed!

**Comment by vpisarev** [comment]
@StevenPuttemans, thank you for another very useful contribution! Let's put it in finally! :+1:

**Comment by StevenPuttemans** [comment]
Thank you!
